File: Sources/HerdrMobile/Transport/Transport.swift
import Foundation

/// The app-side abstraction that executes herdr API requests over SSH.
/// UI code talks to Transport, never to SSH primitives (ADR 0002).
protocol Transport: Sendable {
    /// Verifies the server speaks a protocol version we support and returns
    /// its identity. Must be the first herdr API call on every new connection
    /// path; Host-local session discovery may run before it.
    func ping() async throws -> ServerInfo

    /// Lists the local herdr sessions visible to this SSH account. This is a
    /// Host-level capability and does not depend on the currently selected
    /// API socket, so onboarding can recover from a stale manual selection.
    func listSessions() async throws -> [HerdrSession]

    /// Lists the Agents herdr has detected across all workspaces.
    func listAgents() async throws -> [Agent]

    /// The full session tree in one call: agents plus the workspace context
    /// (labels, worktrees) that `listAgents()` lacks. The Console's snapshot
    /// source (#8) — re-fetched on every events-session `.connected`.
    func sessionSnapshot() async throws -> SessionSnapshot

    /// Reads a Pane's recent terminal output for the Console card snippet.
    func readPane(_ params: PaneReadParams) async throws -> PaneReadResult

    /// Starts a new Agent: the new-agent flow (#12, User Story 8 — dispatch
    /// work from the road). Creates a fresh herdr tab in the chosen workspace,
    /// starts the requested agent in its root pane, and returns the Agent once
    /// the server acknowledges. The new pane also surfaces in the
    /// Console through the normal snapshot/delta machinery (a membership
    /// event triggers a re-snapshot), so callers do not thread the return
    /// value into the list themselves.
    func startAgent(_ request: AgentLaunchRequest) async throws -> Agent

    /// Closes a Pane (`pane.close`): the Agent detail screen's destructive
    /// close action (#13, User Story 9 — a Done agent must not be destroyed
    /// by a stray swipe, so the UI gates this behind an explicit
    /// confirmation). herdr removes the pane and its agent everywhere; the
    /// removal surfaces in the Console through the normal snapshot/delta
    /// machinery (a `pane.closed` membership event triggers a re-snapshot),
    /// so callers do not prune the list themselves. Targeted by the Pane's
    /// id; returns once the server acknowledges.
    func closePane(_ params: PaneTarget) async throws

    /// Opens this Host's dedicated long-lived events channel and subscribes.
    /// Returns once the server acknowledges the subscription; the stream then
    /// carries events in canonical naming until `end()` closes the channel
    /// explicitly. One events channel per Host: a second call while one is
    /// live throws `.eventsChannelAlreadyOpen`.
    ///
    /// Subscribing does not replay existing state (verified against herdr
    /// 0.7.4): sync initial state with `listAgents()` alongside subscribing.
    func subscribeToEvents(_ subscriptions: [EventSubscription]) async throws -> HerdrEventStream

    /// Opens this Host's dedicated terminal channel as a full interactive
    /// Attach: a PTY running `herdr agent attach`, raw bytes both ways until
    /// `end()` closes the channel explicitly. One terminal channel is allowed
    /// per Host, so a second call while one is live throws
    /// `.terminalChannelAlreadyOpen`.
    func attachTerminal(_ request: TerminalAttachRequest) async throws -> TerminalAttachSession

    /// Stages one normalized app-owned image in private Host temporary
    /// storage. Concrete transports own destination selection, restrictive
    /// permissions, partial-file handling, and atomic completion (ADR 0006).
    func stageImage(
        _ image: PreparedImage,
        progress: @escaping @Sendable (ImageStageProgress) async -> Void
    ) async throws -> StagedImage

    /// Reads the Notification Registration file (v1, `plugin/README.md`)
    /// from the herdr-mobile plugin's config dir on this Host; nil when no
    /// device has registered yet. Throws
    /// `NotificationRegistrationError.pluginNotInstalled` when the plugin is
    /// absent, so the ceremony can tell "install the plugin" apart from a
    /// broken read (#72).
    func readNotificationRegistration() async throws -> Data?

    /// Atomically replaces the Notification Registration file with
    /// `contents` (temp file + rename per the v1 contract), creating it when
    /// absent. Same plugin gate as the read.
    func replaceNotificationRegistration(_ contents: Data) async throws

    /// Whether the underlying connection to the Host is still alive. The
    /// reconnect machinery (#18) decides "re-subscribe on this connection or
    /// re-establish it" from this flag.
    var isConnected: Bool { get async }

    /// Tears the connection down explicitly, ending every channel it
    /// carries. Terminal: a closed Transport is not reusable.
    func close() async throws
}

extension Transport {
    /// Test doubles and alternative transports that do not expose Host-level
    /// session discovery can opt out without inventing sessions.
    func listSessions() async throws -> [HerdrSession] { [] }

    /// Non-SSH test doubles and alternative transports can state that SFTP is
    /// unavailable without importing or emulating Citadel.
    func stageImage(
        _ image: PreparedImage,
        progress: @escaping @Sendable (ImageStageProgress) async -> Void
    ) async throws -> StagedImage {
        throw ImageStagingError.sftpUnavailable
    }

    /// Test doubles and alternative transports without a Host-side plugin
    /// can report its absence without emulating the plugin CLI.
    func readNotificationRegistration() async throws -> Data? {
        throw NotificationRegistrationError.pluginNotInstalled
    }

    func replaceNotificationRegistration(_ contents: Data) async throws {
        throw NotificationRegistrationError.pluginNotInstalled
    }
}

/// App-domain request for launching a fresh coding agent.
///
/// herdr protocol 17 split the old topology-changing `agent.start` into
/// `tab.create` followed by a pane-targeted `agent.start`. Keeping that wire
/// choreography behind `Transport` prevents UI code from depending on the
/// server's transport-level request shapes.
struct AgentLaunchRequest: Sendable, Equatable {
    let kind: String
    let name: String
    let arguments: [String]
    let workspaceID: String?

    init(kind: String, name: String, arguments: [String] = [], workspaceID: String? = nil) {
        self.kind = kind
        self.name = name
        self.arguments = arguments
        self.workspaceID = workspaceID
    }
}

/// herdr server identity as reported by `ping`.
struct ServerInfo: Sendable, Equatable {
    let version: String
    let protocolVersion: Int

    init(version: String, protocolVersion: Int) {
        self.version = version
        self.protocolVersion = protocolVersion
    }
}

/// One entry from `herdr session list --json` on a Host.
struct HerdrSession: Sendable, Equatable, Decodable {
    let name: String
    let isDefault: Bool
    let isRunning: Bool

    init(name: String, isDefault: Bool, isRunning: Bool) {
        self.name = name
        self.isDefault = isDefault
        self.isRunning = isRunning
    }

    private enum CodingKeys: String, CodingKey {
        case name
        case isDefault = "default"
        case isRunning = "running"
    }
}

/// The grammar enforced by herdr 0.7.4 for named sessions. Keeping it at the
/// transport boundary prevents malformed discovery output from becoming part
/// of a remote socket path; forms reuse it for immediate feedback.
enum HerdrSessionName {
    static let maximumUTF8Length = 64

    static func isValid(_ name: String) -> Bool {
        guard !name.isEmpty, name != ".", name != ".." else { return false }
        guard name.utf8.count <= maximumUTF8Length else { return false }
        return name.utf8.allSatisfy { byte in
            (0x30...0x39).contains(byte) || (0x41...0x5A).contains(byte)
                || (0x61...0x7A).contains(byte)
                || byte == 0x2E || byte == 0x5F || byte == 0x2D
        }
    }
}

/// Paths passed through the Host's login shell use the conservative quoting
/// subset shared by POSIX shells and fish. Spaces are safe inside single
/// quotes; quote, backslash, and control characters are refused because their
/// single-quote behavior differs across those shells.
enum RemoteShellPath {
    static func quotedAbsolute(_ path: String) -> String? {
        guard path.hasPrefix("/") else { return nil }
        guard path.unicodeScalars.allSatisfy(isQuotable) else { return nil }
        return "'\(path)'"
    }

    static func isQuotableAbsolute(_ path: String) -> Bool {
        quotedAbsolute(path) != nil
    }

    private static func isQuotable(_ scalar: Unicode.Scalar) -> Bool {
        scalar.value >= 0x20 && scalar.value != 0x7F
            && scalar.value != 0x27 && scalar.value != 0x5C
    }
}

/// A coding agent process running inside a herdr Pane.
///
/// The domain view of the generated wire type `AgentInfo`: only the fields
/// the app consumes, with wire-level optionality resolved. `AgentStatus` is
/// the generated raw-string wrapper; Blocked drives sort order and (later)
/// notifications.
struct Agent: Sendable, Equatable {
    let terminalID: String
    /// The agent program herdr detected: "claude", "codex", ... Behavior
    /// stays keyed off this; labels prefer `displayName`.
    let kind: String
    /// The server-reported agent name the herdr TUI shows (`display_agent`,
    /// falling back to `name`); nil when the server reports neither.
    let name: String?
    /// Terminal title with spinner/status glyphs stripped.
    let title: String
    /// Mutable: the Console applies `pane.agent_status_changed` deltas in
    /// place between snapshots.
    var status: AgentStatus
    let workspaceID: String
    let tabID: String
    /// The Pane address used for per-pane subscriptions and attach.
    let paneID: String
    let cwd: String
    let revision: Int

    /// The card's primary label (#41): the server-reported name when present,
    /// otherwise the detected kind.
    var displayName: String { name ?? kind }

    init(
        terminalID: String, kind: String, title: String, status: AgentStatus,
        workspaceID: String, tabID: String, paneID: String, cwd: String, revision: Int,
        name: String? = nil
    ) {
        self.terminalID = terminalID
        self.kind = kind
        self.name = name
        self.title = title
        self.status = status
        self.workspaceID = workspaceID
        self.tabID = tabID
        self.paneID = paneID
        self.cwd = cwd
        self.revision = revision
    }

    /// Maps the generated wire type onto the domain view. Wire-optional
    /// fields degrade instead of failing: herdr's API has no stability
    /// guarantee, and a missing title must not drop the Agent from the list.
    init(_ info: AgentInfo) {
        self.init(
            terminalID: info.terminalID,
            kind: info.agent ?? "unknown",
            title: info.terminalTitleStripped ?? info.terminalTitle ?? "",
            status: info.agentStatus,
            workspaceID: info.workspaceID,
            tabID: info.tabID,
            paneID: info.paneID,
            cwd: info.cwd ?? "",
            revision: info.revision,
            name: Self.nonEmpty(info.displayAgent) ?? Self.nonEmpty(info.name)
        )
    }

    /// An empty wire string carries no name; treating it as missing keeps the
    /// fallback chain from rendering a blank card label.
    private static func nonEmpty(_ value: String?) -> String? {
        value.flatMap { $0.isEmpty ? nil : $0 }
    }
}

/// Where the herdr API socket lives on a Host. Home-relative locations are
/// resolved against the remote home directory, which the Transport resolves
/// over exec once per Host and caches.
enum HerdrSocketLocation: Sendable, Equatable {
    /// The default herdr session: `~/.config/herdr/herdr.sock`.
    case defaultSession
    /// A named session: `~/.config/herdr/sessions/<name>/herdr.sock`.
    case namedSession(String)
    /// An absolute path known in advance; needs no remote resolution.
    case absolutePath(String)

    /// The absolute socket path, given the Host's home directory.
    func path(homeDirectory: String) -> String {
        let home =
            homeDirectory.hasSuffix("/") ? String(homeDirectory.dropLast()) : homeDirectory
        switch self {
        case .defaultSession:
            return "\(home)/.config/herdr/herdr.sock"
        case .namedSession(let name):
            return "\(home)/.config/herdr/sessions/\(name)/herdr.sock"
        case .absolutePath(let path):
            return path
        }
    }
}

/// Transport-level failures: a closed taxonomy so every screen maps errors to
/// user guidance consistently instead of string-matching.
enum TransportError: Error, Sendable, Equatable {
    /// The SSH server could not be reached: connection refused, no route,
    /// or the connection died before authentication.
    case sshUnreachable(detail: String)
    /// The Host rejected our credentials (key not authorized, wrong
    /// password, or the offered auth method is unavailable).
    case authenticationFailed
    /// The device's stored Ed25519 private key cannot be decoded. Reconnecting
    /// cannot repair it; the user must explicitly replace the Device Key.
    case deviceKeyCorrupt
    /// First connect to an unknown Host and the user declined its key
    /// fingerprint; nothing was stored.
    case hostKeyRejected(presented: HostKeyFingerprint)
    /// The Host presented a key that differs from the trusted fingerprint —
    /// possibly a man-in-the-middle. Hard failure; the stored fingerprint is
    /// left untouched.
    case hostKeyMismatch(known: HostKeyFingerprint, presented: HostKeyFingerprint)
    /// The herdr API socket path does not exist on the Host: herdr is not
    /// installed there, or the socket path is wrong.
    case socketNotFound(path: String)
    /// The socket file exists but nothing accepts connections: the herdr
    /// server is not running (cold-start wake is #6).
    case serverNotRunning(path: String)
    /// socat is not at its configured absolute path on the Host.
    case socatMissing(path: String)
    /// The server speaks a herdr protocol version this build does not support.
    case protocolVersionMismatch(server: Int, supported: Int)
    /// The remote home directory could not be resolved, so a home-relative
    /// socket location has no path.
    case homeDirectoryUnresolvable(detail: String)
    /// A second events channel was requested while one is live; each Host
    /// keeps exactly one dedicated events channel (ADR 0002 headroom).
    case eventsChannelAlreadyOpen
    /// A second terminal channel was requested while one is live; each Host
    /// keeps exactly one interactive terminal surface at a time.
    case terminalChannelAlreadyOpen
    /// The request exceeded its per-request deadline; its exec channel was
    /// closed.
    case timedOut
    /// The request's task was cancelled before completing; any exec channel
    /// it held was closed.
    case cancelled
    /// The channel produced bytes that do not decode as a herdr response.
    case malformedResponse(String)
    /// The exec channel failed outside the known failure shapes; carries the
    /// underlying description for diagnostics.
    case channelFailed(detail: String)

    /// Whether reconnecting without user intervention can plausibly recover.
    /// Configuration, trust, authentication, and protocol failures instead
    /// stop so the UI can explain the required action.
    var isRetryable: Bool {
        switch self {
        case .sshUnreachable, .serverNotRunning, .timedOut, .cancelled, .channelFailed:
            true
        case .authenticationFailed, .deviceKeyCorrupt, .hostKeyRejected, .hostKeyMismatch,
            .socketNotFound, .socatMissing, .protocolVersionMismatch,
            .homeDirectoryUnresolvable, .eventsChannelAlreadyOpen,
            .terminalChannelAlreadyOpen, .malformedResponse:
            false
        }
    }
}

/// An error returned by the herdr server inside a response envelope.
struct HerdrAPIError: Error, Sendable, Equatable {
    /// Normalized to a string; the wire schema promises `{"code","message"}`
    /// without pinning the code's JSON type.
    let code: String
    let message: String
}

File: android/core/src/main/java/dev/bybee/heeler/core/transport/Transport.kt
package dev.bybee.heeler.core.transport

import dev.bybee.heeler.core.notifications.NotificationRegistrationError
import dev.bybee.heeler.core.wire.AgentInfo
import dev.bybee.heeler.core.wire.AgentReadParams
import dev.bybee.heeler.core.wire.AgentRenameParams
import dev.bybee.heeler.core.wire.AgentPromptParams
import dev.bybee.heeler.core.wire.AgentSendKeysParams
import dev.bybee.heeler.core.wire.AgentStatus
import dev.bybee.heeler.core.wire.PaneReadParams
import dev.bybee.heeler.core.wire.PaneReadResult
import dev.bybee.heeler.core.wire.PaneTarget
import dev.bybee.heeler.core.wire.SessionSnapshot
import dev.bybee.heeler.core.wire.WorkspaceRenameParams
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.io.File

/**
 * The app-side abstraction that executes herdr API requests over SSH. UI code
 * talks to Transport, never to SSH primitives (ADR 0011).
 */
interface Transport {
    /**
     * Verifies a supported protocol and returns server identity. This must be
     * the first herdr API call on every new connection path; Host-local session
     * discovery may run before it.
     */
    suspend fun ping(): ServerInfo

    /**
     * Lists herdr sessions visible to this SSH account. This is Host-level and
     * does not depend on the selected API socket, so onboarding can recover
     * from a stale manual selection.
     */
    suspend fun listSessions(): List<HerdrSession> = emptyList()

    /** Lists Agents herdr detected across all workspaces. */
    suspend fun listAgents(): List<Agent>

    /**
     * Lists supported Agent kinds whose canonical executables are available on
     * this Host. Alternative transports report no detected kinds by default.
     */
    suspend fun availableAgentKinds(): List<SupportedAgentKind> = emptyList()

    /**
     * The full session tree in one call: agents plus workspace labels and
     * worktrees. The Console snapshot source, re-fetched on every events
     * session connection.
     */
    suspend fun sessionSnapshot(): SessionSnapshot

    /** Reads a Pane's recent terminal output for a Console card snippet. */
    suspend fun readPane(params: PaneReadParams): PaneReadResult

    /**
     * Reads an Agent's terminal output. Unlike `pane.read`, this preserves
     * alternate-screen history semantics: history-capable sources fail honestly
     * while an Agent works instead of silently degrading to visible screen.
     */
    suspend fun readAgent(params: AgentReadParams): PaneReadResult

    /**
     * Delivers one complete local draft through `agent.prompt`. It deliberately
     * omits `wait`: the response acknowledges delivery into the Agent pane;
     * Agent Status events report subsequent work.
     */
    suspend fun promptAgent(params: AgentPromptParams): Agent

    /** Sends control keys to an Agent (`agent.send_keys`). */
    suspend fun sendAgentKeys(params: AgentSendKeysParams)

    /**
     * Starts a fresh Agent by creating a tab in the selected workspace, then
     * starting the requested kind in its root Pane. Membership surfaces through
     * normal snapshot/delta machinery, so callers do not mutate lists directly.
     */
    suspend fun startAgent(request: AgentLaunchRequest): Agent

    /**
     * Starts an Agent in a fresh git worktree. `worktree.create` resolves from
     * the source workspace cwd and returns a workspace whose root Pane already
     * runs a shell, so this skips `tab.create`; the `agent_pane_busy` readiness
     * retry still applies.
     */
    suspend fun startAgentInNewWorktree(request: AgentLaunchRequest, worktree: WorktreeSpec): Agent

    /** Closes a Pane. Its membership removal reaches callers through snapshots/events. */
    suspend fun closePane(params: PaneTarget)

    /**
     * Renames an Agent. A null name clears its custom name; herdr rejects a
     * non-null value outside `^[a-z][a-z0-9_-]{0,31}$` with
     * `invalid_agent_name`. Rename events do not carry the new agent name, so
     * consumers re-snapshot after this call rather than mutating local state.
     */
    suspend fun renameAgent(params: AgentRenameParams)

    /**
     * Renames a workspace. herdr accepts any label, including whitespace or
     * empty labels; callers rely on `workspace.renamed` rather than local edits.
     */
    suspend fun renameWorkspace(params: WorkspaceRenameParams)

    /**
     * Opens the Host's one dedicated long-lived events channel. `end()` closes
     * it explicitly. Replayed events on subscription are ordinary change
     * signals, never state: callers still fetch a snapshot alongside it.
     */
    suspend fun subscribeToEvents(subscriptions: List<EventSubscription>): HerdrEventStream

    /**
     * Opens the Host's dedicated PTY running `herdr agent attach`. Exactly one
     * terminal channel is allowed per Host.
     */
    suspend fun attachTerminal(request: TerminalAttachRequest): TerminalAttachSession

    /**
     * Stages one normalized app-owned image in private Host temporary storage.
     * Concrete transports own destination selection, restrictive permissions,
     * partial-file handling, and atomic completion (ADR 0006).
     */
    suspend fun stageImage(
        image: PreparedImage,
        progress: suspend (AttachmentStageProgress) -> Unit,
    ): StagedImage = throw AttachmentStagingError.SftpUnavailable

    /** Stages one app-owned file under the same SFTP/permission/atomic policy. */
    suspend fun stageFile(
        file: PreparedFile,
        progress: suspend (AttachmentStageProgress) -> Unit,
    ): StagedFile = throw AttachmentStagingError.SftpUnavailable

    /**
     * Reads the Heeler plugin's notification registration file, returning null
     * when no device has registered. Plugin absence throws
     * [NotificationRegistrationError.PluginNotInstalled] rather than masking
     * installation state as a broken read.
     */
    suspend fun readNotificationRegistration(): ByteArray? =
        throw NotificationRegistrationError.PluginNotInstalled

    /** Atomically replaces the plugin registration file (temp + rename). */
    suspend fun replaceNotificationRegistration(contents: ByteArray) {
        throw NotificationRegistrationError.PluginNotInstalled
    }

    /** Reads the plugin's optional `notify.json` config file. */
    suspend fun readNotificationConfig(): ByteArray? =
        throw NotificationRegistrationError.PluginNotInstalled

    /** Atomically replaces the plugin's optional `notify.json` config file. */
    suspend fun replaceNotificationConfig(contents: ByteArray) {
        throw NotificationRegistrationError.PluginNotInstalled
    }

    /**
     * Lists custom slash commands for an Agent kind. SSH implementations probe
     * the remote filesystem; alternative transports report no skills by default.
     */
    suspend fun listSkills(query: SkillListQuery): List<AgentSkill> = emptyList()

    /** Reads one reported skill document in full, subject to the remote cap. */
    suspend fun readSkillFile(path: String): String =
        throw TransportError.ChannelFailed("This transport cannot read skill files.")

    /** Whether the underlying Host connection is alive. */
    suspend fun isConnected(): Boolean

    /** Explicit terminal teardown. A closed Transport is not reusable. */
    suspend fun close()
}

/** The interactive Agent kinds supported by herdr protocol 17. */
enum class SupportedAgentKind(
    val rawValue: String,
    val displayName: String,
    val executable: String = rawValue,
) {
    PI("pi", "Pi"),
    CLAUDE("claude", "Claude Code"),
    CODEX("codex", "Codex"),
    GEMINI("gemini", "Gemini CLI"),
    CURSOR("cursor", "Cursor Agent", "cursor-agent"),
    DEVIN("devin", "Devin CLI"),
    ANTIGRAVITY("agy", "Antigravity"),
    CLINE("cline", "Cline"),
    OMP("omp", "OMP"),
    MASTRACODE("mastracode", "Mastra Code"),
    OPENCODE("opencode", "OpenCode"),
    COPILOT("copilot", "GitHub Copilot CLI"),
    KIMI("kimi", "Kimi CLI"),
    KIRO("kiro", "Kiro CLI", "kiro-cli"),
    DROID("droid", "Droid"),
    AMP("amp", "Amp"),
    GROK("grok", "Grok Build"),
    HERMES("hermes", "Hermes Agent"),
    KILO("kilo", "Kilo Code"),
    QODERCLI("qodercli", "Qoder CLI"),
    MAKI("maki", "Maki"),
    ;

    companion object {
        fun fromRawValue(rawValue: String): SupportedAgentKind? =
            entries.firstOrNull { it.rawValue == rawValue }
    }
}

/**
 * App-domain request for launching a fresh coding Agent.
 *
 * herdr protocol 17 split the old topology-changing `agent.start` into
 * `tab.create` followed by pane-targeted `agent.start`. Keeping that wire
 * choreography behind [Transport] prevents UI code from depending on the
 * server's transport-level request shapes.
 */
data class AgentLaunchRequest(
    val kind: String,
    val name: String,
    val arguments: List<String> = emptyList(),
    val workspaceID: String? = null,
    /**
     * Working directory for the fresh tab, carried when a launch starts from
     * another Agent screen and should land in the same place. Null lets herdr
     * fall back to the workspace directory.
     */
    val cwd: String? = null,
)

/**
 * What the skills probe needs: the Agent kind and its launch project root.
 * The root is the worktree checkout or Agent start directory, deliberately
 * not the foreground cwd: an Agent changing directories cannot alter skills.
 */
data class SkillListQuery(
    val kind: SupportedAgentKind,
    /** Null skips project skills when the Agent's project is unknown. */
    val projectRoot: String? = null,
)

/**
 * App-domain refinements for a fresh-worktree launch. Null fields use herdr's
 * defaults: a generated `worktree/<name>` branch from HEAD in herdr's root.
 */
data class WorktreeSpec(val branch: String? = null, val base: String? = null)

/**
 * herdr server identity reported by `ping`. A protocol newer than this
 * generated snapshot remains usable; newer additive features are advisory.
 */
data class ServerInfo(
    val version: String,
    val protocolVersion: Int,
    /** Advisory only: this Host is newer than the schema snapshot. */
    val exceedsGeneratedProtocol: Boolean = false,
)

/** One entry from `herdr session list --json` on a Host. */
@Serializable
data class HerdrSession(
    val name: String,
    @SerialName("default") val isDefault: Boolean,
    @SerialName("running") val isRunning: Boolean,
)

/**
 * The grammar herdr 0.7.4 enforces for named sessions. Keeping it at the
 * transport boundary prevents malformed discovery output from entering a
 * remote socket path; forms reuse it for immediate feedback.
 */
object HerdrSessionName {
    const val MAXIMUM_UTF8_LENGTH = 64

    fun isValid(name: String): Boolean =
        name.isNotEmpty() && name != "." && name != ".." &&
            name.encodeToByteArray().size <= MAXIMUM_UTF8_LENGTH &&
            name.all { character ->
                character in '0'..'9' || character in 'A'..'Z' ||
                    character in 'a'..'z' || character == '.' ||
                    character == '_' || character == '-'
            }
}

/**
 * Paths interpolated into the Host login shell use the conservative quoting
 * subset shared by POSIX shells and fish. Spaces are safe in single quotes;
 * quote, backslash, and control characters are refused because their quote
 * behavior differs across those shells.
 */
object RemoteShellPath {
    fun quotedAbsolute(path: String): String? =
        path.takeIf(::isQuotableAbsolute)?.let { "'$it'" }

    fun isQuotableAbsolute(path: String): Boolean =
        path.startsWith('/') && path.all { scalar ->
            scalar.code >= 0x20 && scalar.code != 0x7f && scalar != '\'' && scalar != '\\'
        }
}

/** A coding Agent process running inside a herdr Pane. */
data class Agent(
    val terminalID: String,
    val kind: String,
    val name: String?,
    val title: String,
    var status: AgentStatus,
    val workspaceID: String,
    val tabID: String,
    val paneID: String,
    val cwd: String,
    val revision: Int,
) {
    /** The server name when present, otherwise the detected kind. */
    val displayName: String get() = name ?: kind

    companion object {
        /**
         * Maps a generated [AgentInfo] into the domain view. Wire-optional
         * fields degrade instead of failing: herdr has no stability guarantee,
         * and a missing title must not drop an Agent from the list.
         */
        fun fromWire(info: AgentInfo): Agent = Agent(
            terminalID = info.terminalID,
            kind = info.agent ?: "unknown",
            name = info.displayAgent.nonEmpty() ?: info.name.nonEmpty(),
            title = info.terminalTitleStripped ?: info.terminalTitle ?: "",
            status = info.agentStatus,
            workspaceID = info.workspaceID,
            tabID = info.tabID,
            paneID = info.paneID,
            cwd = info.cwd.orEmpty(),
            revision = info.revision,
        )
    }
}

/** Where the herdr API socket lives on a Host. */
sealed interface HerdrSocketLocation {
    /** The default `~/.config/herdr/herdr.sock`. */
    data object DefaultSession : HerdrSocketLocation

    /** The named `~/.config/herdr/sessions/<name>/herdr.sock`. */
    data class NamedSession(val name: String) : HerdrSocketLocation {
        init {
            require(HerdrSessionName.isValid(name)) { "Invalid herdr session name." }
        }
    }

    /** An absolute path known in advance and needing no remote-home resolution. */
    data class AbsolutePath(val path: String) : HerdrSocketLocation

    /** Resolves a home-relative location to an absolute socket path. */
    fun path(homeDirectory: String): String {
        val home = homeDirectory.removeSuffix("/")
        return when (this) {
            DefaultSession -> "$home/.config/herdr/herdr.sock"
            is NamedSession -> "$home/.config/herdr/sessions/$name/herdr.sock"
            is AbsolutePath -> path
        }
    }
}

/** App-owned normalized image ready for transport; local naming carries no source metadata. */
data class PreparedImage(
    val file: File,
    val format: PreparedImageFormat,
    val pixelWidth: Int,
    val pixelHeight: Int,
    val byteCount: Long,
) {
    companion object {
        const val MAXIMUM_ENCODED_BYTE_COUNT = 16 * 1024 * 1024
    }
}

enum class PreparedImageFormat(val fileExtension: String) {
    JPEG("jpg"),
    PNG("png"),
}

/** App-owned file ready for transport, with a safe extension only. */
data class PreparedFile(
    val file: File,
    val fileExtension: String,
    val byteCount: Long,
) {
    val remoteFilename: String
        get() {
            val extension = safeExtension(fileExtension)
            return if (extension.isEmpty()) "file" else "file.$extension"
        }

    companion object {
        const val MAXIMUM_BYTE_COUNT = 64 * 1024 * 1024

        fun safeExtension(candidate: String): String {
            val normalized = candidate.lowercase()
            return normalized.takeIf { value ->
                value.isNotEmpty() && value.length <= 16 && value.all { it in 'a'..'z' || it in '0'..'9' }
            }.orEmpty()
        }
    }
}

/** Upload progress with a clamped display fraction. */
data class AttachmentStageProgress(val transferredBytes: Long, val totalBytes: Long) {
    val fractionCompleted: Double
        get() = if (totalBytes <= 0L) 0.0 else
            (transferredBytes.toDouble() / totalBytes.toDouble()).coerceIn(0.0, 1.0)
}

class StagedImage(path: String) {
    val path: String = checkedStagedHostPath(path)
    override fun equals(other: Any?): Boolean = other is StagedImage && path == other.path
    override fun hashCode(): Int = path.hashCode()
}

class StagedFile(path: String) {
    val path: String = checkedStagedHostPath(path)
    override fun equals(other: Any?): Boolean = other is StagedFile && path == other.path
    override fun hashCode(): Int = path.hashCode()
}

/** Attachment staging failures with retry semantics matching the iOS transport. */
sealed class AttachmentStagingError(message: String) : Exception(message) {
    data object InvalidRemotePath : AttachmentStagingError("Invalid remote path.")
    data object InvalidPreparedSource : AttachmentStagingError("Invalid prepared source.")
    data object LocalReadFailed : AttachmentStagingError("Prepared source could not be read.")
    data object RemoteTemporaryDirectoryFailed : AttachmentStagingError("Remote temporary directory failed.")
    data object SftpUnavailable : AttachmentStagingError("SFTP is unavailable.")
    data object PermissionEnforcementFailed : AttachmentStagingError("Remote permissions could not be enforced.")
    data object ByteCountMismatch : AttachmentStagingError("Uploaded byte count differs.")
    data object TransferFailed : AttachmentStagingError("Attachment transfer failed.")
    data object Cancelled : AttachmentStagingError("Attachment transfer was cancelled.")

    val isRetryable: Boolean
        get() = this === TransferFailed || this === Cancelled
}

private fun String?.nonEmpty(): String? = this?.takeIf(String::isNotEmpty)

private fun checkedStagedHostPath(path: String): String {
    if (!path.startsWith('/') || path == "/" || path.any { it.code < 0x20 || it.code == 0x7f }) {
        throw AttachmentStagingError.InvalidRemotePath
    }
    return path
}

/** Generated wire aliases exposed through the transport boundary. */
typealias SessionSnapshotResult = SessionSnapshot
typealias PaneReadRequest = PaneReadParams
typealias PaneReadResponse = PaneReadResult
typealias AgentReadRequest = AgentReadParams
typealias AgentPromptRequest = AgentPromptParams
typealias AgentRenameRequest = AgentRenameParams
typealias WorkspaceRenameRequest = WorkspaceRenameParams
